### PR TITLE
[shard] make ShardedTensor a torch.Tensor subclass

### PR DESCRIPTION
**Pull request by wanchaol**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #77592

This is the reland of PR https://github.com/pytorch/pytorch/pull/74695, which was reverted due to some internal failures.

It also removes the ShardedTensorInterface change, we will delay that
change later if we found there's a need to do that.

### CODE REVIEW

**Comment by wanchaol** [comment]
@wanchaol has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D36423976).

**Comment by wanchaol** [comment]
@wanchaol has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D36423976).

File: torch/distributed/_shard/sharded_tensor/api.py [hunk 1]
```diff
@@ -115,11 +114,47 @@ class ShardedTensor(object):
         individual GPU, via ``torch.cuda.set_device()``
 
     """
-
-    def __new__(cls, *args, **kwargs):
-        # Use __new__ for logging purposes.
+    _sharding_spec: shard_spec.ShardingSpec
+    _metadata: ShardedTensorMetadata
+
```
Comment (pritamdamania87) [inline torch/distributed/_shard/sharded_tensor/api.py:119]: nit: Do we need slots?

File: torch/distributed/_shard/sharded_tensor/api.py [hunk 2]
```diff
@@ -790,183 +827,12 @@ def local_shards(self) -> List[Shard]:
...
-        if dim < -len(size) or dim >= len(size):
-            raise ValueError(
-                "Argument ``dim`` must be within the range of tensor "
-                f"dimensions [-{len(size)}, {len(size)})"
-            )
-        return size[dim]
-
-
-    def is_pinned(self) -> bool:
+    def is_pinned(self) -> bool:  # type: ignore[override]
```
Comment (pritamdamania87) [inline torch/distributed/_shard/sharded_tensor/api.py:830]: Can't this be registered as a default tensor op too?

**Review by pritamdamania87 — APPROVED** [review]

File: torch/distributed/_shard/sharded_tensor/api.py (hunk 2 above)
Comment (wanchaol) [reply torch/distributed/_shard/sharded_tensor/api.py:830]: So the issue with `is_pinned` is that it's actually getting dispatch to C++ and then to `torch_dispatch`, if we want to register it the way as default op, we will need to change `torch_dispatch` to handle this call in a special manner. I can think of how to do this in a follow up PR.

**Pull request closed without merging**